src/app/services/connectivity/connectivity.service.ts

```diff
@@ -1,30 +1,51 @@
 import { Injectable } from '@angular/core';
-import { BehaviorSubject, Observable } from 'rxjs';
+
+import { ConnectableObservable, Observable } from 'rxjs';
+import { multicast, publishReplay } from 'rxjs/operators';
+
 
 @Injectable({
   providedIn: 'root'
 })
 export class ConnectivityService {
-  private monitor: Observable<boolean>;
-  private subject: BehaviorSubject<boolean>;
-
-  constructor() {
-    this.monitor = new Observable((connectivityObserver) => {
-      // map 'offline' and 'online' window events to boolean 'connnectivity state'
-      window.addEventListener('offline', () => {
-        connectivityObserver.next(false);
-      });
-      window.addEventListener('online', () => {
-        connectivityObserver.next(true);
-      });
-    });
-    // obtain initial connectivity state from navigator.onLine property
-    this.subject = new BehaviorSubject(navigator.onLine);
-    this.monitor.subscribe(this.subject);
+  public static readonly OFFLINE_EVENT = 'offline';
+  public static readonly ONLINE_EVENT = 'online';
+  private connectivityMonitor: Observable<boolean>;
+  private connectivityObserver: ConnectableObservable<boolean>;
+
+  constructor() { }
+
+  initConnectivityObserver() {
+    this.connectivityMonitor = new Observable(
+      (connectivityStateObserver) => {
+        const offlineEventListener = () => {
+          connectivityStateObserver.next(false); // Browser offline -> connectivity: false
+        };
+        const onlineEventListener = () => {
+          connectivityStateObserver.next(true); // Browser online -> connectivity: true
+        };
+
+        window.addEventListener(ConnectivityService.OFFLINE_EVENT, offlineEventListener);
+        window.addEventListener(ConnectivityService.ONLINE_EVENT, onlineEventListener);
+
+        // Emit the current connectivity
+        connectivityStateObserver.next(navigator.onLine);
+        return () => {
+          window.removeEventListener(ConnectivityService.OFFLINE_EVENT, offlineEventListener);
+          window.removeEventListener(ConnectivityService.ONLINE_EVENT, onlineEventListener);
+        };
+      }
+    );
+
+    // Publish a multicast ReplaySubject that emits the most recent connectivity status
+    this.connectivityObserver = this.connectivityMonitor.pipe(publishReplay(1)) as ConnectableObservable<boolean>;
+    this.connectivityObserver.connect();
```

Comment:
Is this line necessary? If you expose one instance of this Observable from the service, isn't that sufficient, without having to invoke connect or refCount?

`this.isConnected$ = this.connectivityMonitor.pipe(shareReplay());`

I forget the exact differences between share and publish replay, but I thought they wrap the functionality you're adding here?

I haven't done it this way before so not 100%.

Reply:
The aim is for lazy initialisation and the ability to reconnect. I was under the impression that `refCount()` would re-connect (and re-start the underlying subscription). Though now I'm looking around to confirm and it seems I may be mistaken... seems the behaviour may have changed (or broke) with rxjs 5. What's your preferred source for reliable info in these situations?

Admittedly it's probably YAGNI for this website but I was curious about the multicast pattern and making this service re-usable in more complex apps.